dds/DCPS/BuiltInTopicUtils.cpp

```diff
@@ -34,6 +38,418 @@ const char* const BUILT_IN_CONNECTION_RECORD_TOPIC_TYPE = "CONNECTION_RECORD_BUI
+  ACE_UNUSED_ARG(timestamp);
+  return DDS::HANDLE_NIL;
+#endif /* DDS_HAS_MINIMUM_BIT */
+}
+
+void BitSubscriber::remove_thread_status(const InternalThreadBuiltinTopicData& ts)
+{
+#ifndef DDS_HAS_MINIMUM_BIT
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+
+  if (!bit_subscriber_) {
+    return;
+  }
+
+  DDS::DataReader_var d = bit_subscriber_->lookup_datareader(BUILT_IN_INTERNAL_THREAD_TOPIC);
+
+  if (!d) {
+    return;
+  }
+
+  InternalThreadBuiltinTopicDataDataReaderImpl* bit = dynamic_cast<InternalThreadBuiltinTopicDataDataReaderImpl*>(d.in());
+
+  bit->set_instance_state(bit->lookup_instance(ts), DDS::NOT_ALIVE_DISPOSED_INSTANCE_STATE);
+#else
+  ACE_UNUSED_ARG(ts);
+#endif
+}
+
+void BitSubscriber::remove_participant_i(DDS::InstanceHandle_t part_ih)
+{
+#ifndef DDS_HAS_MINIMUM_BIT
+  if (part_ih != DDS::HANDLE_NIL) {
+    DDS::DataReader_var d = bit_subscriber_->lookup_datareader(BUILT_IN_PARTICIPANT_TOPIC);
+    if (!d) {
+      return;
+    }
+
+    ParticipantBuiltinTopicDataDataReaderImpl* bit = dynamic_cast<ParticipantBuiltinTopicDataDataReaderImpl*>(d.in());
+
+    bit->set_instance_state(part_ih, DDS::NOT_ALIVE_DISPOSED_INSTANCE_STATE);
+  }
+#else
+  ACE_UNUSED_ARG(part_ih);
+#endif
+}
+
+void BitSubscriber::remove_participant_location_i(DDS::InstanceHandle_t loc_ih)
+{
+#ifndef DDS_HAS_MINIMUM_BIT
+  if (loc_ih != DDS::HANDLE_NIL) {
+    DDS::DataReader_var d = bit_subscriber_->lookup_datareader(BUILT_IN_PARTICIPANT_LOCATION_TOPIC);
+    if (!d) {
+      return;
+    }
+
+    ParticipantLocationBuiltinTopicDataDataReaderImpl* bit = dynamic_cast<ParticipantLocationBuiltinTopicDataDataReaderImpl*>(d.in());
+
+    bit->set_instance_state(loc_ih, DDS::NOT_ALIVE_DISPOSED_INSTANCE_STATE);
+  }
+#else
+  ACE_UNUSED_ARG(loc_ih);
+#endif
+}
+
+void BitSubscriber::bit_pub_listener_hack(DomainParticipantImpl* participant)
+{
+#ifndef DDS_HAS_MINIMUM_BIT
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+
+  if (!bit_subscriber_) {
+    return;
+  }
+
+  DDS::DataReader_var dr =
+    bit_subscriber_->lookup_datareader(BUILT_IN_PUBLICATION_TOPIC);
+  DDS::PublicationBuiltinTopicDataDataReader_var bit_pub_dr =
+    DDS::PublicationBuiltinTopicDataDataReader::_narrow(dr);
+
+  if (!CORBA::is_nil(bit_pub_dr.in())) {
+    DDS::DataReaderListener_var listener = bit_pub_dr->get_listener();
+    if (CORBA::is_nil(listener.in())) {
+      DDS::DataReaderListener_var bit_pub_listener =
+        new BitPubListenerImpl(participant);
+      bit_pub_dr->set_listener(bit_pub_listener, DDS::DATA_AVAILABLE_STATUS);
+      // Must call on_data_available when attaching a listener late - samples may be waiting
+      SubscriberImpl* sub = dynamic_cast<SubscriberImpl*>(bit_subscriber_.in());
+      DataReaderImpl* reader = dynamic_cast<DataReaderImpl*>(bit_pub_dr.in());
+      TheServiceParticipant->job_queue()->enqueue(make_rch<DataReaderImpl::OnDataAvailable>(rchandle_from(sub), bit_pub_listener, rchandle_from(reader), true, false, false));
```

Comment:
Does the `OnDataAvailable` constructor need both the reader and the subscriber?  Each reader already knows which subscriber it belongs to.